juju/application.py

```diff
@@ -641,48 +643,85 @@ async def refresh(
+            resolved_charm_with_channel_results = await charms_facade.ResolveCharms(resolve=[client.ResolveCharmWithChannel(
+                charm_origin=origin,
+                switch_charm=False,
+                reference=charm_url,
+            )])
+            resolved_charm = resolved_charm_with_channel_results.results[0]
+
+            if resolved_charm.error is not None:
+                err = resolved_charm.error
+                raise JujuError(f'{err.code} : {err.message}')
+            dest_origin = resolved_charm.charm_origin
+            charm_url = resolved_charm.url
 
-        if switch is not None:
-            charm_url = switch
-            if not charm_url.startswith('cs:'):
-                charm_url = 'cs:' + charm_url
         else:
-            charm_url = self.data['charm-url']
-            charm_url = charm_url.rpartition('-')[0]
-            if revision is not None:
-                charm_url = "%s-%d" % (charm_url, revision)
+            charms_facade = client.CharmsFacade.from_connection(self.connection)
+            charmstore = self.model.charmstore
+            charmstore_entity = None
+            if switch is not None:
+                charm_url = switch
+                if not charm_url.startswith('cs:'):
+                    charm_url = 'cs:' + charm_url
             else:
-                charmstore_entity = await charmstore.entity(charm_url,
-                                                            channel=channel)
-                charm_url = charmstore_entity['Id']
-
-        if charm_url == self.data['charm-url']:
-            raise JujuError('already running charm "%s"' % charm_url)
-
-        # TODO (caner) : this needs to be revisited and updated with the charmhub stuff
-        origin = client.CharmOrigin(source="charm-store",
-                                    risk=channel,
-                                    )
-        # Update charm
-        await charms_facade.AddCharm(
-            url=charm_url,
-            force=force,
-            charm_origin=origin,
-        )
+                charm_url = self.data['charm-url']
+                charm_url = charm_url.rpartition('-')[0]
+                if revision is not None:
+                    charm_url = "%s-%d" % (charm_url, revision)
+                else:
+                    charmstore_entity = await charmstore.entity(charm_url, channel=channel)
+                    charm_url = charmstore_entity['Id']
 
-        # Update resources
-        if not charmstore_entity:
-            charmstore_entity = await charmstore.entity(charm_url,
-                                                        channel=channel)
-        store_resources = charmstore_entity['Meta']['resources']
+            if charm_url == self.data['charm-url']:
+                raise JujuError('already running charm "%s"' % charm_url)
 
+            dest_origin = client.CharmOrigin(source="charm-store", risk=channel)
```

Comment:
I'm curious what the charm-origin in the db for this application looks like after AddCharm/SetCharm are called below. Thinking of the bug I just put a pr up to fix: LP1990182